Show loaded modules under '/status'

ac000 opened the pull request:
This pull-request comprises five commits with the overall goal of showing what language modules Unit has loaded _and_ from where, under the `/status` control API endpoint.

E.g

```json
{
    "modules": {
        "python": [
            {
                "version": "3.12.3",
                "lib": "/opt/unit/modules/python.unit.so"
            },
            {
                "version": "3.12.1",
                "lib": "/opt/unit/modules/python-3.12.1.unit.so"
            }
        ],
        "wasm": {
            "version": "0.1",
            "lib": "/opt/unit/modules/wasm.unit.so"
        },
        "wasm-wasi-component": {
            "version": "0.1",
            "lib": "/opt/unit/modules/wasm_wasi_component.unit.so"
        }
    },
...
}
```

This could be a useful debugging aid.

Patches

- status: Constify a bunch of local variables

Preparatory patch that `static const`'s a bunch of nxt_string()'s before I add some more...

- status: Use a variable to represent the status member index

Preparatory patch to use a variable to track the top-level object's index in the JSON rather than hard coding it.

- Flow the language module name into nxt_app_lang_module_t

This stores the language module name in the nxt_app_lang_module_t strcuture.

- status: Show list of loaded language modules

Main patch that adds a new top-level `modules` section.

- tests: Fix `/status' endpoint tests for new 'modules' section

See commit messages for more information...

ac000 commented:
Fix build on Clang and older GCC

```
$ git range-diff 300f3cb5...74bea8c4
1:  300f3cb5 ! 1:  74bea8c4 status: Show list of loaded language modules
    @@ src/nxt_status.c
     +nxt_inline const nxt_str_t *
     +nxt_status_app_type2name(nxt_app_type_t type)
     +{
    -+#define nxt_str_sconst(s)  ((nxt_str_t *)&(static const nxt_str_t)nxt_string(s))
    ++#define nxt_string_const(s)  (&(const nxt_str_t)nxt_string(s))
     +
     +    switch (type) {
     +    case NXT_APP_EXTERNAL:
    -+        return nxt_str_sconst("external");
    ++        return nxt_string_const("external");
     +    case NXT_APP_PYTHON:
    -+        return nxt_str_sconst("python");
    ++        return nxt_string_const("python");
     +    case NXT_APP_PHP:
    -+        return nxt_str_sconst("php");
    ++        return nxt_string_const("php");
     +    case NXT_APP_PERL:
    -+        return nxt_str_sconst("perl");
    ++        return nxt_string_const("perl");
     +    case NXT_APP_RUBY:
    -+        return nxt_str_sconst("ruby");
    ++        return nxt_string_const("ruby");
     +    case NXT_APP_JAVA:
    -+        return nxt_str_sconst("java");
    ++        return nxt_string_const("java");
     +    case NXT_APP_WASM:
    -+        return nxt_str_sconst("wasm");
    ++        return nxt_string_const("wasm");
     +    case NXT_APP_WASM_WC:
    -+        return nxt_str_sconst("wasm-wasi-component");
    ++        return nxt_string_const("wasm-wasi-component");
     +    case NXT_APP_UNKNOWN:
    -+        return nxt_str_sconst("unknown");
    ++        return nxt_string_const("unknown");
     +    }
     +
     +    /* Avoid 'warning: control reaches end of non-void function' */
     +    nxt_unreachable();
     +
    -+#undef nxt_str_sconst
    ++#undef nxt_string_const
     +}
      
      
```

ac000 commented:
@andrey-zelenkov 

This patch set which adds a new `modules` section to the `/status` endpoint e.g

```json
{
    "modules": {
        "python": {
            "version": "3.12.2",
            "lib": "/opt/unit/modules/python.unit.so"
        },
        "wasm": {
            "version": "0.1",
            "lib": "/opt/unit/modules/wasm.unit.so"
        },
        "wasm-wasi-component": {
            "version": "0.1",
            "lib": "/opt/unit/modules/wasm_wasi_component.unit.so"
        }
    },
...
}
```

Broke the pytests `/status` checks

```
INTERNALERROR> Traceback (most recent call last):
INTERNALERROR>   File "/usr/lib/python3.12/site-packages/_pytest/main.py", line 269, in wrap_session
INTERNALERROR>     config.hook.pytest_sessionstart(session=session)
INTERNALERROR>   File "/usr/lib/python3.12/site-packages/pluggy/_hooks.py", line 493, in __call__
INTERNALERROR>     return self._hookexec(self.name, self._hookimpls, kwargs, firstresult)
INTERNALERROR>            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
INTERNALERROR>   File "/usr/lib/python3.12/site-packages/pluggy/_manager.py", line 115, in _hookexec
INTERNALERROR>     return self._inner_hookexec(hook_name, methods, kwargs, firstresult)
INTERNALERROR>            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
INTERNALERROR>   File "/usr/lib/python3.12/site-packages/pluggy/_callers.py", line 152, in _multicall
INTERNALERROR>     return outcome.get_result()
INTERNALERROR>            ^^^^^^^^^^^^^^^^^^^^
INTERNALERROR>   File "/usr/lib/python3.12/site-packages/pluggy/_result.py", line 114, in get_result
INTERNALERROR>     raise exc.with_traceback(exc.__traceback__)
INTERNALERROR>   File "/usr/lib/python3.12/site-packages/pluggy/_callers.py", line 77, in _multicall
INTERNALERROR>     res = hook_impl.function(*args)
INTERNALERROR>           ^^^^^^^^^^^^^^^^^^^^^^^^^
INTERNALERROR>   File "/home/andrew/src/unit/test/conftest.py", line 168, in pytest_sessionstart
INTERNALERROR>     unit = unit_run()
INTERNALERROR>            ^^^^^^^^^^
INTERNALERROR>   File "/home/andrew/src/unit/test/conftest.py", line 339, in unit_run
INTERNALERROR>     Status._check_zeros()
INTERNALERROR>   File "/home/andrew/src/unit/test/unit/status.py", line 9, in _check_zeros
INTERNALERROR>     assert Status.control.conf_get('/status') == {
INTERNALERROR>            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
INTERNALERROR> AssertionError
```

Due to the new `modules` section missing from `./test/unit/status.py`. I tried adding it in

```diff
diff --git ./test/unit/status.py ./test/unit/status.py
index 95096a96..f103f561 100644
--- ./test/unit/status.py
+++ ./test/unit/status.py
@@ -7,6 +7,7 @@ class Status:
 
     def _check_zeros():
         assert Status.control.conf_get('/status') == {
+            'modules': {},
             'connections': {
                 'accepted': 0,
                 'active': 0,
```

But, doh!, obviously it won't be empty, there will be at _least_ the python language module, there may be other modules depending on what people had configured and built.

These modules of course also won't be of a fixed version or path...

Thinking about how to fix this, in order of preference/maintainability etc

1) Simply remove this check (`_check_zeros()`)

This will become evermore harder to keep working as we add more persistent dynamic content to this endpoint, think start times etc.

2) Ignore the `modules` section in the check either completely or with some wildcard regex of the `modules` object content.

3) Dynamically create a suitable `modules` object depending on the current environment.

I'm currently not sure about (2) and don't know if (3) is even feasible...

Idea's?

ac000 commented:
- Drop the pytests patch
- Use a statement expression in `nxt_status_app_type2name()` to do what I attempted to do initially. On first looks it seems like it might not work as it looks like it will use the same variable for each switch case, but the compiler generates unique variables by appending .0, .1 etc to each...

```
$ git range-diff 691ba993...3cb718f0
1:  74bea8c4 ! 1:  3cb718f0 status: Show list of loaded language modules
    @@ src/nxt_status.c
     +nxt_inline const nxt_str_t *
     +nxt_status_app_type2name(nxt_app_type_t type)
     +{
    -+#define nxt_string_const(s)  (&(const nxt_str_t)nxt_string(s))
    ++#define nxt_str_sconst(s)                                                     \
    ++    ({                                                                        \
    ++        static const nxt_str_t _scstr = nxt_string(s);                        \
    ++        &_scstr;                                                              \
    ++     })
     +
     +    switch (type) {
     +    case NXT_APP_EXTERNAL:
    -+        return nxt_string_const("external");
    ++        return nxt_str_sconst("external");
     +    case NXT_APP_PYTHON:
    -+        return nxt_string_const("python");
    ++        return nxt_str_sconst("python");
     +    case NXT_APP_PHP:
    -+        return nxt_string_const("php");
    ++        return nxt_str_sconst("php");
     +    case NXT_APP_PERL:
    -+        return nxt_string_const("perl");
    ++        return nxt_str_sconst("perl");
     +    case NXT_APP_RUBY:
    -+        return nxt_string_const("ruby");
    ++        return nxt_str_sconst("ruby");
     +    case NXT_APP_JAVA:
    -+        return nxt_string_const("java");
    ++        return nxt_str_sconst("java");
     +    case NXT_APP_WASM:
    -+        return nxt_string_const("wasm");
    ++        return nxt_str_sconst("wasm");
     +    case NXT_APP_WASM_WC:
    -+        return nxt_string_const("wasm-wasi-component");
    ++        return nxt_str_sconst("wasm-wasi-component");
     +    case NXT_APP_UNKNOWN:
    -+        return nxt_string_const("unknown");
    ++        return nxt_str_sconst("unknown");
     +    }
     +
     +    /* Avoid 'warning: control reaches end of non-void function' */
     +    nxt_unreachable();
     +
    -+#undef nxt_string_const
    ++#undef nxt_str_sconst
     +}
      
      
2:  691ba993 < -:  -------- test: Update '/status' tests for new 'modules' section
```

andrey-zelenkov commented:
>```
>"python": {
>           "version": "3.12.2",
>            "lib": "/opt/unit/modules/python.unit.so"
>        },
> ```

Side note: it could be an array since multiple versions of the same module are allowed.

> Thinking about how to fix this, in order of preference/maintainability etc

For now, you can remove `modules` key from tested object (`resp = Status.control.conf_get('/status') ... del reps['modules']`) or, better yet, just test the desired metrics. Something like:

```
diff --git a/test/unit/status.py b/test/unit/status.py
--- a/test/unit/status.py
+++ b/test/unit/status.py
@@ -6,16 +6,16 @@ class Status:
     control = Control()
 
     def _check_zeros():
-        assert Status.control.conf_get('/status') == {
-            'connections': {
+        status = Status.control.conf_get('/status')
+
+        assert status['connections'] == {
                 'accepted': 0,
                 'active': 0,
                 'idle': 0,
                 'closed': 0,
-            },
-            'requests': {'total': 0},
-            'applications': {},
         }
+        assert status['requests'] == {'total': 0}
+        assert status['applications'] == {}
 
     def init(status=None):
         Status._status = (
```

Reminder for the future: after implementing `modules` it will be necessary to rework `test/unit/check/discover_available.py`.

ac000 commented:
@andrey-zelenkov 

Thanks for the patch and while it works OK on master

```
$ pytest test/test_status.py
which: no go in (/home/andrew/.local/bin:/home/andrew/bin:/usr/share/Modules/bin:/usr/local/bin:/usr/local/sbin:/usr/bin:/usr/sbin)
which: no go in (/home/andrew/.local/bin:/home/andrew/bin:/usr/share/Modules/bin:/usr/local/bin:/usr/local/sbin:/usr/bin:/usr/sbin)
============================================= test session starts ==============================================
platform linux -- Python 3.12.3, pytest-7.4.3, pluggy-1.3.0 -- /usr/bin/python3
cachedir: .pytest_cache
rootdir: /home/andrew/src/unit/test
configfile: pytest.ini
collected 5 items                                                                                              

test/test_status.py::test_status PASSED
test/test_status.py::test_status_requests PASSED
test/test_status.py::test_status_connections PASSED
test/test_status.py::test_status_applications PASSED
test/test_status.py::test_status_proxy PASSED

============================================== 5 passed in 8.35s ===============================================
```

With this it fails...

```
$ pytest test/test_status.py
which: no go in (/home/andrew/.local/bin:/home/andrew/bin:/usr/share/Modules/bin:/usr/local/bin:/usr/local/sbin:/usr/bin:/usr/sbin)
which: no go in (/home/andrew/.local/bin:/home/andrew/bin:/usr/share/Modules/bin:/usr/local/bin:/usr/local/sbin:/usr/bin:/usr/sbin)
============================================= test session starts ==============================================
platform linux -- Python 3.12.3, pytest-7.4.3, pluggy-1.3.0 -- /usr/bin/python3
cachedir: .pytest_cache
rootdir: /home/andrew/src/unit/test
configfile: pytest.ini
collected 5 items                                                                                              

test/test_status.py::test_status PASSED
test/test_status.py::test_status_requests FAILEDPath to unit.log:
/tmp/unit-test-gu6mskrx/unit.log


test/test_status.py::test_status_connections FAILEDPath to unit.log:
/tmp/unit-test-gu6mskrx/unit.log


test/test_status.py::test_status_applications FAILEDPath to unit.log:
/tmp/unit-test-gu6mskrx/unit.log


test/test_status.py::test_status_proxy FAILEDPath to unit.log:
/tmp/unit-test-gu6mskrx/unit.log

2024/07/02 20:07:08 [info] 341931#341931 "empty" prototype started
2024/07/02 20:07:08 [info] 341932#341932 "empty" application started


=================================================== FAILURES ===================================================
_____________________________________________ test_status_requests _____________________________________________

skip_alert = <function skip_alert.<locals>._skip at 0x7ffadbfb6700>

    def test_status_requests(skip_alert):
        skip_alert(r'Python failed to import module "blah"')
    
        assert 'success' in client.conf(
            {
                "listeners": {
                    "*:8080": {"pass": "routes"},
                    "*:8081": {"pass": "applications/empty"},
                    "*:8082": {"pass": "applications/blah"},
                },
                "routes": [{"action": {"return": 200}}],
                "applications": {
                    "empty": app_default(),
                    "blah": {
                        "type": client.get_application_type(),
                        "processes": {"spare": 0},
                        "module": "blah",
                    },
                },
            },
        )
    
        Status.init()
    
        assert client.get()['status'] == 200
>       assert Status.get('/requests/total') == 1, '2xx'

test/test_status.py:61: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
test/unit/status.py:40: in get
    diff = Status.diff()
test/unit/status.py:36: in diff
    return find_diffs(Status.control.conf_get('/status'), Status._status)
test/unit/status.py:29: in find_diffs
    k: find_diffs(d1.get(k, 0), d2.get(k, 0))
test/unit/status.py:29: in find_diffs
    k: find_diffs(d1.get(k, 0), d2.get(k, 0))
test/unit/status.py:29: in find_diffs
    k: find_diffs(d1.get(k, 0), d2.get(k, 0))
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

d1 = '3.12.3', d2 = '3.12.3'

    def find_diffs(d1, d2):
        if isinstance(d1, dict) and isinstance(d2, dict):
            return {
                k: find_diffs(d1.get(k, 0), d2.get(k, 0))
                for k in d1
                if k in d2
            }
    
>       return d1 - d2
E       TypeError: unsupported operand type(s) for -: 'str' and 'str'

test/unit/status.py:34: TypeError
___________________________________________ test_status_connections ____________________________________________

    def test_status_connections():
        assert 'success' in client.conf(
            {
                "listeners": {
                    "*:8080": {"pass": "routes"},
                    "*:8081": {"pass": "applications/delayed"},
                },
                "routes": [{"action": {"return": 200}}],
                "applications": {
                    "delayed": app_default("delayed"),
                },
            },
        )
    
        Status.init()
    
        # accepted, closed
    
        assert client.get()['status'] == 200
>       check_connections(1, 0, 0, 1)

test/test_status.py:116: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
test/test_status.py:13: in check_connections
    assert Status.get('/connections') == {
test/unit/status.py:40: in get
    diff = Status.diff()
test/unit/status.py:36: in diff
    return find_diffs(Status.control.conf_get('/status'), Status._status)
test/unit/status.py:29: in find_diffs
    k: find_diffs(d1.get(k, 0), d2.get(k, 0))
test/unit/status.py:29: in find_diffs
    k: find_diffs(d1.get(k, 0), d2.get(k, 0))
test/unit/status.py:29: in find_diffs
    k: find_diffs(d1.get(k, 0), d2.get(k, 0))
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

d1 = '3.12.3', d2 = '3.12.3'

    def find_diffs(d1, d2):
        if isinstance(d1, dict) and isinstance(d2, dict):
            return {
                k: find_diffs(d1.get(k, 0), d2.get(k, 0))
                for k in d1
                if k in d2
            }
    
>       return d1 - d2
E       TypeError: unsupported operand type(s) for -: 'str' and 'str'

test/unit/status.py:34: TypeError
___________________________________________ test_status_applications ___________________________________________

    def test_status_applications():
        def check_applications(expert):
            apps = list(client.conf_get('/status/applications').keys()).sort()
            assert apps == expert.sort()
    
        def check_application(name, running, starting, idle, active):
            assert Status.get(f'/applications/{name}') == {
                'processes': {
                    'running': running,
                    'starting': starting,
                    'idle': idle,
                },
                'requests': {'active': active},
            }
    
        client.load('delayed')
        Status.init()
    
        check_applications(['delayed'])
>       check_application('delayed', 0, 0, 0, 0)

test/test_status.py:168: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
test/test_status.py:155: in check_application
    assert Status.get(f'/applications/{name}') == {
test/unit/status.py:40: in get
    diff = Status.diff()
test/unit/status.py:36: in diff
    return find_diffs(Status.control.conf_get('/status'), Status._status)
test/unit/status.py:29: in find_diffs
    k: find_diffs(d1.get(k, 0), d2.get(k, 0))
test/unit/status.py:29: in find_diffs
    k: find_diffs(d1.get(k, 0), d2.get(k, 0))
test/unit/status.py:29: in find_diffs
    k: find_diffs(d1.get(k, 0), d2.get(k, 0))
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

d1 = '3.12.3', d2 = '3.12.3'

    def find_diffs(d1, d2):
        if isinstance(d1, dict) and isinstance(d2, dict):
            return {
                k: find_diffs(d1.get(k, 0), d2.get(k, 0))
                for k in d1
                if k in d2
            }
    
>       return d1 - d2
E       TypeError: unsupported operand type(s) for -: 'str' and 'str'

test/unit/status.py:34: TypeError
______________________________________________ test_status_proxy _______________________________________________

    def test_status_proxy():
        assert 'success' in client.conf(
            {
                "listeners": {
                    "*:8080": {"pass": "routes"},
                    "*:8081": {"pass": "applications/empty"},
                },
                "routes": [
                    {
                        "match": {"uri": "/"},
                        "action": {"proxy": "http://127.0.0.1:8081"},
                    }
                ],
                "applications": {
                    "empty": app_default(),
                },
            },
        )
    
        Status.init()
    
        assert client.get()['status'] == 200
>       check_connections(2, 0, 0, 2)

test/test_status.py:241: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
test/test_status.py:13: in check_connections
    assert Status.get('/connections') == {
test/unit/status.py:40: in get
    diff = Status.diff()
test/unit/status.py:36: in diff
    return find_diffs(Status.control.conf_get('/status'), Status._status)
test/unit/status.py:29: in find_diffs
    k: find_diffs(d1.get(k, 0), d2.get(k, 0))
test/unit/status.py:29: in find_diffs
    k: find_diffs(d1.get(k, 0), d2.get(k, 0))
test/unit/status.py:29: in find_diffs
    k: find_diffs(d1.get(k, 0), d2.get(k, 0))
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

d1 = '3.12.3', d2 = '3.12.3'

    def find_diffs(d1, d2):
        if isinstance(d1, dict) and isinstance(d2, dict):
            return {
                k: find_diffs(d1.get(k, 0), d2.get(k, 0))
                for k in d1
                if k in d2
            }
    
>       return d1 - d2
E       TypeError: unsupported operand type(s) for -: 'str' and 'str'

test/unit/status.py:34: TypeError
=========================================== short test summary info ============================================
FAILED test/test_status.py::test_status_requests - TypeError: unsupported operand type(s) for -: 'str' and 'str'
FAILED test/test_status.py::test_status_connections - TypeError: unsupported operand type(s) for -: 'str' and 'str'
FAILED test/test_status.py::test_status_applications - TypeError: unsupported operand type(s) for -: 'str' and 'str'
FAILED test/test_status.py::test_status_proxy - TypeError: unsupported operand type(s) for -: 'str' and 'str'
========================================= 4 failed, 1 passed in 0.81s ==========================================
```

ac000 commented:
This patch (on top of the previous one) seems to get things going again...

```diff
diff --git ./test/unit/status.py ./test/unit/status.py
index 2a95022d..50c21c42 100644
--- ./test/unit/status.py
+++ ./test/unit/status.py
@@ -31,6 +31,9 @@ class Status:
                     if k in d2
                 }
 
+            if isinstance(d1, str):
+                return d1 == d2
+
             return d1 - d2
 
         return find_diffs(Status.control.conf_get('/status'), Status._status)
```

It's enough to just check d1 as d1 and d2 seem to always be the same type...

ac000 commented:
- Some commit message tweaks
- A fix for the `test_status.py` pytest (incl patch from @andrey-zelenkov)

```
$ git range-diff 3cb718f0...2aa8779a
1:  8bdc2889 ! 1:  e73bef7a status: Constify a bunch of local variables
    @@ Commit message
         status: Constify a bunch of local variables
     
         This is yet more missed constification, due in this case to me searching
    -    for 'static nxt_str_t  ` but these only having a single space after the
    +    for 'static nxt_str_t  ' but these only having a single space after the
         type...
     
         Anyway no problem, this can be a preparatory patch for adding further
2:  b6d76abb = 2:  bb6cdcf0 status: Use a variable to represent the status member index
3:  3cb718f0 ! 3:  d02d9e70 status: Show list of loaded language modules
    @@ Commit message
         This can be useful for debugging to show exactly what modules Unit has
         loaded _and_ from where.
     
    +    In cases where there are multiple modules of the same type but different
    +    version, they will each be listed individually.
    +
         Due to the module name not actually being kept around (it could be added
         but would be a much more intrusive patch) we need to map module types to
         their name.
-:  -------- > 4:  2aa8779a tests: Fix `/status' endpoint tests for new 'modules' section
```

hongzhidao commented on src/nxt_status.c:
```diff
@@ -6,18 +6,60 @@
 
 nxt_conf_value_t *
-nxt_status_get(nxt_status_report_t *report, nxt_mp_t *mp)
+nxt_status_get(nxt_task_t *task, nxt_status_report_t *report, nxt_mp_t *mp)
```
Good stuff.

We could get the job of getting the module name done earlier.
The modules/languages are discovered in the `discover` process which exits after finishing its job.
I feel the module name can be set in it.
1. set module name in `nxt_discovery_module()`.
2. then it passed to the main process, `nxt_main_port_modules_handler()`.
    I feel it makes sense to add a field to save the module name in `nxt_app_lang_module_t`.
3. use the module name directly in the status API.

Btw, you can get thread with `nxt_thread()` without adding `task` parameter.

ac000 replied:
Sure, this is what I wrote in the commit message...

```
Due to the module name not actually being kept around (it could be added
but would be a much more intrusive patch) we need to map module types to
their name.
```

Do you want the more intrusive patch?

ac000 replied:
Hmm, do JSON objects at the same level have to have unique names?...

hongzhidao replied:
Oh, I didn't notice you had the idea.
I feel it's not more intrusive. Before the module name wasn't used, now it looks good to be added before control and router processes are created.
But it's just an idea, how good depends on the code.

hongzhidao replied:
Oh, I didn't notice you had the idea.
I feel it's not more intrusive. Before the module name wasn't used, now it looks good to be added before control and router processes are created.
It just intuitively makes more sense, how good depends on the code.

ac000 replied:
OK, so it's not _that_ bad... <https://git.digital-domain.net/unit.git/commit/?h=status-modules&id=d03b0d2cf431eceb3dd5ace97460749a1f1096f9>

hongzhidao commented:
> Side note: it could be an array since multiple versions of the same module are allowed.

Yep, for example, users can run python 2.6 and python 2.7 together.

ac000 commented:
> > Side note: it could be an array since multiple versions of the same module are allowed.
> 
> Yep, for example, users can run python 2.6 and python 2.7 together.

That's mentioned in the commit message, they will each get their own entry.

ac000 commented:
> > > Side note: it could be an array since multiple versions of the same module are allowed.
> > 
> > 
> > Yep, for example, users can run python 2.6 and python 2.7 together.
> 
> That's mentioned in the commit message, they will each get their own entry.

Hmm, do JSON objects at the same level have to have unique names?...

hongzhidao commented:
> > > > Side note: it could be an array since multiple versions of the same module are allowed.
> > > 
> > > 
> > > Yep, for example, users can run python 2.6 and python 2.7 together.
> > 
> > 
> > That's mentioned in the commit message, they will each get their own entry.
> 
> Hmm, do JSON objects at the same level have to have unique names?...

Yep, the object property name is unique.

ac000 commented:
- Use nxt_thread() instead of passing the task structure around
-  Flow the language module name through and remove the type2name mapping function
- Handle multiple versions of the same language module

```
$ git range-diff --creation-factor=80 2aa8779a...e9a63926
-:  -------- > 1:  d03b0d2c Flow the language module name into nxt_app_lang_module_t
1:  d02d9e70 ! 2:  70708bd7 status: Show list of loaded language modules
    @@ Commit message
     
           {
               "modules": {
    -              "python": {
    -                  "version": "3.12.2",
    -                  "lib": "/opt/unit/modules/python.unit.so"
    -              },
    +              "python": [
    +                  {
    +                      "version": "3.12.3",
    +                       "lib": "/opt/unit/modules/python.unit.so"
    +                  },
    +                  {
    +                      "version": "3.12.1",
    +                      "lib": "/opt/unit/modules/python-3.12.1.unit.so"
    +                  }
    +              ],
     
                   "wasm": {
                       "version": "0.1",
    @@ Commit message
         This can be useful for debugging to show exactly what modules Unit has
         loaded _and_ from where.
     
    -    In cases where there are multiple modules of the same type but different
    -    version, they will each be listed individually.
    -
    -    Due to the module name not actually being kept around (it could be added
    -    but would be a much more intrusive patch) we need to map module types to
    -    their name.
    -
    -    I have done this using a static inline function with a switch()
    -    statement rather say an array, so that if the nxt_app_type_t enum is
    -    ever updated and the switch statement here isn't, the compiler will
    -    shout at you.
    -
         Closes: https://github.com/nginx/unit/issues/1343
         Signed-off-by: Andrew Clayton <a.clayton@nginx.com>
     
    - ## src/nxt_controller.c ##
    -@@ src/nxt_controller.c: nxt_controller_status_handler(nxt_task_t *task, nxt_port_recv_msg_t *msg,
    -     req = data;
    - 
    -     if (msg->port_msg.type == NXT_PORT_MSG_RPC_READY) {
    --        status = nxt_status_get((nxt_status_report_t *) msg->buf->mem.pos,
    -+        status = nxt_status_get(task,
    -+                                (nxt_status_report_t *) msg->buf->mem.pos,
    -                                 req->conn->mem_pool);
    -     } else {
    -         status = NULL;
    -
      ## src/nxt_status.c ##
     @@
      #include <nxt_main.h>
      #include <nxt_conf.h>
      #include <nxt_status.h>
     +#include <nxt_application.h>
    -+
    -+
    -+nxt_inline const nxt_str_t *
    -+nxt_status_app_type2name(nxt_app_type_t type)
    -+{
    -+#define nxt_str_sconst(s)                                                     \
    -+    ({                                                                        \
    -+        static const nxt_str_t _scstr = nxt_string(s);                        \
    -+        &_scstr;                                                              \
    -+     })
    -+
    -+    switch (type) {
    -+    case NXT_APP_EXTERNAL:
    -+        return nxt_str_sconst("external");
    -+    case NXT_APP_PYTHON:
    -+        return nxt_str_sconst("python");
    -+    case NXT_APP_PHP:
    -+        return nxt_str_sconst("php");
    -+    case NXT_APP_PERL:
    -+        return nxt_str_sconst("perl");
    -+    case NXT_APP_RUBY:
    -+        return nxt_str_sconst("ruby");
    -+    case NXT_APP_JAVA:
    -+        return nxt_str_sconst("java");
    -+    case NXT_APP_WASM:
    -+        return nxt_str_sconst("wasm");
    -+    case NXT_APP_WASM_WC:
    -+        return nxt_str_sconst("wasm-wasi-component");
    -+    case NXT_APP_UNKNOWN:
    -+        return nxt_str_sconst("unknown");
    -+    }
    -+
    -+    /* Avoid 'warning: control reaches end of non-void function' */
    -+    nxt_unreachable();
    -+
    -+#undef nxt_str_sconst
    -+}
      
      
      nxt_conf_value_t *
    --nxt_status_get(nxt_status_report_t *report, nxt_mp_t *mp)
    -+nxt_status_get(nxt_task_t *task, nxt_status_report_t *report, nxt_mp_t *mp)
    + nxt_status_get(nxt_status_report_t *report, nxt_mp_t *mp)
      {
    -     size_t            i;
    -     uint32_t          idx = 0;
    -     nxt_str_t         name;
    -     nxt_int_t         ret;
    -+    nxt_array_t       *langs;
    -     nxt_status_app_t  *app;
    +-    size_t            i;
    +-    uint32_t          idx = 0;
    +-    nxt_str_t         name;
    +-    nxt_int_t         ret;
    +-    nxt_status_app_t  *app;
     -    nxt_conf_value_t  *status, *obj, *apps, *app_obj;
    -+    nxt_conf_value_t  *status, *obj, *mods, *apps, *app_obj;
    ++    size_t                 i, nr_langs;
    ++    size_t                 lang_cnts[NXT_APP_UNKNOWN] = { 1 };
    ++    uint32_t               idx = 0;
    ++    nxt_str_t              name;
    ++    nxt_int_t              ret;
    ++    nxt_array_t            *langs;
    ++    nxt_thread_t           *thr;
    ++    nxt_app_type_t         type, prev_type;
    ++    nxt_status_app_t       *app;
    ++    nxt_conf_value_t       *status, *obj, *mods, *apps, *app_obj, *mod_obj;
    ++    nxt_app_lang_module_t  *modules;
      
     +    static const nxt_str_t  modules_str = nxt_string("modules");
     +    static const nxt_str_t  version_str = nxt_string("version");
    @@ src/nxt_status.c: nxt_status_get(nxt_status_report_t *report, nxt_mp_t *mp)
              return NULL;
          }
      
    -+    langs = task->thread->runtime->languages;
    ++    thr = nxt_thread();
    ++    langs = thr->runtime->languages;
     +
    -+    /* Don't reserve space for NXT_APP_EXTERNAL */
    -+    mods = nxt_conf_create_object(mp, langs->nelts - 1);
    ++    modules = langs->elts;
    ++    /*
    ++     * We need to count the number of unique languages to correctly
    ++     * allocate the below mods object.
    ++     *
    ++     * We also need to count how many of each language.
    ++     *
    ++     * Start by skipping past NXT_APP_EXTERNAL which is always the
    ++     * first entry.
    ++     */
    ++    for (i = 1, nr_langs = 0, prev_type = NXT_APP_UNKNOWN; i < langs->nelts;
    ++         i++)
    ++    {
    ++        type = modules[i].type;
    ++
    ++        lang_cnts[type]++;
    ++
    ++        if (type == prev_type) {
    ++            continue;
    ++        }
    ++
    ++        nr_langs++;
    ++        prev_type = type;
    ++    }
    ++
    ++    mods = nxt_conf_create_object(mp, nr_langs);
     +    if (nxt_slow_path(mods == NULL)) {
     +        return NULL;
     +    }
     +
     +    nxt_conf_set_member(status, &modules_str, mods, idx++);
     +
    -+    i = 0;
    -+    for (size_t l = 0; i < langs->nelts; i++) {
    -+        nxt_str_t              item;
    -+        const nxt_str_t        *mod_name;
    -+        nxt_conf_value_t       *mod_obj;
    -+        nxt_app_lang_module_t  *modules = langs->elts;
    ++    i = 1;
    ++    obj = mod_obj = NULL;
    ++    prev_type = NXT_APP_UNKNOWN;
    ++    for (size_t l = 0, a = 0; i < langs->nelts; i++) {
    ++        nxt_str_t  item, mod_name;
     +
    -+        if (modules[i].type == NXT_APP_EXTERNAL) {
    -+            continue;
    ++        type = modules[i].type;
    ++        if (type != prev_type) {
    ++            a = 0;
    ++
    ++            if (lang_cnts[type] == 1) {
    ++                mod_obj = nxt_conf_create_object(mp, 2);
    ++                obj = mod_obj;
    ++            } else {
    ++                mod_obj = nxt_conf_create_array(mp, lang_cnts[type]);
    ++            }
    ++
    ++            if (nxt_slow_path(mod_obj == NULL)) {
    ++                return NULL;
    ++            }
    ++
    ++            mod_name.start = (u_char *)modules[i].name;
    ++            mod_name.length = strlen(modules[i].name);
    ++            nxt_conf_set_member(mods, &mod_name, mod_obj, l++);
     +        }
     +
    -+        mod_obj = nxt_conf_create_object(mp, 2);
    -+        if (nxt_slow_path(mod_obj == NULL)) {
    -+            return NULL;
    ++        if (lang_cnts[type] > 1) {
    ++            obj = nxt_conf_create_object(mp, 2);
    ++            nxt_conf_set_element(mod_obj, a++, obj);
     +        }
     +
    -+        mod_name = nxt_status_app_type2name(modules[i].type);
    -+
    -+        nxt_conf_set_member(mods, mod_name, mod_obj, l++);
    -+
     +        item.start = modules[i].version;
     +        item.length = nxt_strlen(modules[i].version);
    -+        nxt_conf_set_member_string(mod_obj, &version_str, &item, 0);
    ++        nxt_conf_set_member_string(obj, &version_str, &item, 0);
     +
    -+        item.start = (u_char *) modules[i].file;
    ++        item.start = (u_char *)modules[i].file;
     +        item.length = strlen(modules[i].file);
    -+        nxt_conf_set_member_string(mod_obj, &lib_str, &item, 1);
    ++        nxt_conf_set_member_string(obj, &lib_str, &item, 1);
    ++
    ++        prev_type = type;
     +    }
     +
          obj = nxt_conf_create_object(mp, 4);
          if (nxt_slow_path(obj == NULL)) {
              return NULL;
    -
    - ## src/nxt_status.h ##
    -@@ src/nxt_status.h: typedef struct {
    - } nxt_status_report_t;
    - 
    - 
    --nxt_conf_value_t *nxt_status_get(nxt_status_report_t *report, nxt_mp_t *mp);
    -+nxt_conf_value_t *nxt_status_get(nxt_task_t *task, nxt_status_report_t *report,
    -+     nxt_mp_t *mp);
    - 
    - 
    - #endif /* _NXT_STATUS_H_INCLUDED_ */
2:  2aa8779a = 3:  e9a63926 tests: Fix `/status' endpoint tests for new 'modules' section
```

ac000 commented:
- Use a `uint16_t` for the language module counts array. I can't see there ever being more than 64K versions of a language module being loaded and this saves 48 bytes of stack space.

```
$ git range-diff e9a63926...9f0c9f3d
1:  70708bd7 ! 1:  f393a199 status: Show list of loaded language modules
    @@ src/nxt_status.c
     -    nxt_status_app_t  *app;
     -    nxt_conf_value_t  *status, *obj, *apps, *app_obj;
     +    size_t                 i, nr_langs;
    -+    size_t                 lang_cnts[NXT_APP_UNKNOWN] = { 1 };
    ++    uint16_t               lang_cnts[NXT_APP_UNKNOWN] = { 1 };
     +    uint32_t               idx = 0;
     +    nxt_str_t              name;
     +    nxt_int_t              ret;
2:  e9a63926 = 2:  9f0c9f3d tests: Fix `/status' endpoint tests for new 'modules' section
```

hongzhidao reviewed (approved):
LGTM.
Btw, is it necessary to show the corresponding configuration example in the commit?

ac000 commented:
On Wed, 10 Jul 2024 12:41:51 -0700
hongzhidao ***@***.***> wrote:

> @hongzhidao approved this pull request.
> 
> LGTM.
> Btw, is it necessary to show the corresponding configuration example in the commit?

Necessary? No. But it doesn't hurt and shows exactly the intended
behaviour of the patch...

ac000 commented:
Rebased with master

```
$ git range-diff 9f0c9f3d...f4ba4b55
 -:  -------- >  1:  d62a5e2c contrib: updated njs to 0.8.5
 -:  -------- >  2:  a9aa9e76 python: Support application factories
 -:  -------- >  3:  d67d3501 tests: Add tests for python application factories
 -:  -------- >  4:  36213522 Fix certificate deletion for array type certificates
 -:  -------- >  5:  ff6d5045 python: Constify some local static variables
 -:  -------- >  6:  0bd18ecd Packages: clean up EOL debian-based distributions
 -:  -------- >  7:  151305eb Packages: added Ubuntu 24.04 "noble" support
 -:  -------- >  8:  6c04c7dc Packages: don't redefine FORTIFY_SOURCE on Ubuntu
 -:  -------- >  9:  804a12fe Packages: remove support for EOL Fedora versions (35-38)
 -:  -------- > 10:  929a275a Packages: removed CentOS 6 leftover
 -:  -------- > 11:  63fb95ed Packages: removed CentOS 7 due to EOL
 -:  -------- > 12:  0e6cc6dc auto: Fix some indentation in auto/modules/wasm-wasi-component
 -:  -------- > 13:  b5fe3eaf auto, wasm-wc: Copy the .so into build/lib/unit/modules/
 -:  -------- > 14:  706ea1a6 tools/unitctl: Enable Multi Socket Support
 -:  -------- > 15:  6ca27c32 contrib: make sha512sum check compatible with FreeBSD 14+
 1:  e73bef7a = 16:  081e5115 status: Constify a bunch of local variables
 2:  bb6cdcf0 = 17:  c8d70c3f status: Use a variable to represent the status member index
 3:  d03b0d2c = 18:  55041ef9 Flow the language module name into nxt_app_lang_module_t
 4:  f393a199 = 19:  707f4ef8 status: Show list of loaded language modules
 5:  9f0c9f3d = 20:  f4ba4b55 tests: Fix `/status' endpoint tests for new 'modules' section
```
